libraries/reliable_amqp_publisher/reliable_amqp_publisher.cpp

```diff
@@ -0,0 +1,236 @@
+#include <eosio/reliable_amqp_publisher/reliable_amqp_publisher.hpp>
+
+#include <fc/network/url.hpp>
+#include <fc/exception/exception.hpp>
+#include <fc/io/cfile.hpp>
+#include <fc/log/logger_config.hpp> //set_os_thread_name()
+
+#include <amqpcpp.h>
+#include <amqpcpp/libboostasio.h>
+
+#include <thread>
+#include <chrono>
+#include <future>
+#include <algorithm>
+
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/filesystem.hpp>
+
+namespace eosio {
+
+struct reliable_amqp_publisher_callbacks {
+   virtual void amqp_ready() = 0;
+   virtual void amqp_error(const char* msg) = 0;
+};
+
+struct reliable_amqp_publisher_handler : AMQP::LibBoostAsioHandler {
+   reliable_amqp_publisher_handler(reliable_amqp_publisher_callbacks* callbacks, boost::asio::io_context& ctx) :
+      AMQP::LibBoostAsioHandler(ctx),
+      cbs(callbacks) {}
+
+   void onReady(AMQP::TcpConnection* c) override {
+      cbs->amqp_ready();
+   }
+
+   //disable heartbeat for now
+   uint16_t onNegotiate(AMQP::TcpConnection* connection, uint16_t interval) override {
+      return 0;
+   }
+
+   void onError(AMQP::TcpConnection*, const char* message) override {
+      cbs->amqp_error(message);
+   }
+
+   reliable_amqp_publisher_callbacks* cbs;
+};
+
+struct reliable_amqp_publisher_impl final : reliable_amqp_publisher_callbacks {
+   reliable_amqp_publisher_impl(const std::string& url, const std::string& exchange, const boost::filesystem::path& unconfirmed_path) :
+     data_file_path(unconfirmed_path), exchange(exchange) {
+      std::string host = "localhost", user = "guest", pass = "guest", path;
+      uint16_t port = 5672;
+      fc::url parsed_url(url);
+      FC_ASSERT(parsed_url.proto() == "amqp", "Only amqp:// URIs are supported for AMQP addresses");
+      if(parsed_url.host() && parsed_url.host()->size())
+         host = *parsed_url.host();
+      if(parsed_url.user())
+         user = *parsed_url.user();
+      if(parsed_url.pass())
+         pass = *parsed_url.pass();
+      if(parsed_url.port())
+         port = *parsed_url.port();
+      if(parsed_url.path())
+         path = parsed_url.path()->string();
+
+      std::stringstream hostss;
+      hostss << "amqp://" << user << ":" << pass << "@" << host << ":" << port << "/" << path;
+      amqp_address = std::make_unique<AMQP::Address>(hostss.str());
+
+      if(boost::filesystem::exists(data_file_path)) {
+         try {
+            fc::datastream<fc::cfile> file;
+            file.set_file_path(data_file_path);
+            file.open("rb");
+            fc::raw::unpack(file, message_deque);
+         } FC_RETHROW_EXCEPTIONS(error, "Failed to load previously cached AMQP messages from ${f}", ("f", (fc::path)data_file_path));
+      }
+      boost::system::error_code ec;
+      boost::filesystem::remove(data_file_path, ec);
```

Comment:
I think you should verify that you can create the `data_file_path` before starting so that you are less likely to fail on shutdown because of file permissions or invalid path.